blog/not-so-basic-base-r-functions/index.qmd

```diff
@@ -133,52 +129,212 @@ print("Notice the quotation marks?")
+
+## 5. with
+
+If you use base R, you've likely encountered the dollar signs `$` when evaluating expressions with variables from a data frame. The `with()` function lets you reference columns directly, eliminating the need to repeat the data frame name multiple times. This makes your code more concise and easier to read.
+
+::: {.column-margin}
+**Resources**
+
+* [with documentation](https://stat.ethz.ch/R-manual/R-devel/library/base/html/with.html)
+* [When to use 'with' function and why is it good?](https://stackoverflow.com/questions/42283479/when-to-use-with-function-and-why-is-it-good)
+:::
+
+So, instead of writing `plot(mtcars$hp, mtcars$mpg)`, you can write:
+
+```{webr}
+with(mtcars, plot(hp, mpg))
+```
+
+This is particularly handy to use with the base R pipe `|>`:
+
+```{webr}
+mtcars |> with(plot(hp, mpg))
+```
+
+[Michael Love's Tweet](https://twitter.com/mikelove/status/1197190835994943489) shows how to connect a dplyr chain to a base plot function using `with()`:
+
+```{webr}
+webr::install("dplyr")
+library(dplyr)
+        
+mtcars |> dplyr::filter(hp > 100) |> with(plot(hp, mpg))
 ```
 
-I first learned about `coplot()` thanks to Eric Leung's tweet:
+## 6. lengths
 
-<blockquote class="twitter-tweet"><p lang="en" dir="ltr">TIL about coplots in base <a href="https://twitter.com/hashtag/RStats?src=hash&amp;ref_src=twsrc%5Etfw">#RStats</a>. in my nearly decade use of R, i&#39;ve never come across this function to quickly explore pairs of variables conditioned on a third<br><br>library(palmerpenguins)<br>coplot(body_mass_g ~ bill_length_mm|species, data = penguins)<a href="https://t.co/EVpI738VzO">https://t.co/EVpI738VzO</a> <a href="https://t.co/zcTw4HGvnZ">pic.twitter.com/zcTw4HGvnZ</a></p>&mdash; Eric Leung 梁挺亮 (@erictleung) <a href="https://twitter.com/erictleung/status/1555656643056058369?ref_src=twsrc%5Etfw">August 5, 2022</a></blockquote> <script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script> 
+`lengths()` is a more efficient version of `sapply(df, length)`. `length()` determines the number of elements in an object, and `lengths()` will provide the lengths of elements across columns in the data frame.
 
-## 4. split
+::: {.column-margin}
+**Resources**
+
+* [lengths documentation](https://stat.ethz.ch/R-manual/R-devel/library/base/html/lengths.html)
+* [Summing the lengths of lists inside a list in R](https://stackoverflow.com/questions/71969291/summing-the-lengths-of-lists-inside-a-list-in-r)
+:::
+
+```{webr}
+df <- data.frame(A = c(1, 2, 3), B = c(4, 5, 6), C = c(7, 8, 9))
+
+lengths(df)
+```
+
+Pretty straightforward but I think it is a neat function :)
+
+## 7. Null-coalescing operator in R, %||%
+
+OK, this one isn't in base R -- yet! In the [upcoming release](https://developer.r-project.org/blosxom.cgi/R-devel/2023/10/26#n2023-10-26), R will automatically provide the [null-coalescing operator]((https://en.m.wikipedia.org/wiki/Null_coalescing_operator), `%||%`. Per the release notes:
```

Comment:
The link to the null-coalescing operator didn't render, there's an extra opening parenthesis here: `[null-coalescing operator]((https://en.m.wikipedia.org/wiki/Null_coalescing_operator)`